service/history/workflow/mutable_state_impl.go

```diff
@@ -1186,6 +1186,19 @@ func (ms *MutableStateImpl) GetWorkflowCloseTime(ctx context.Context) (time.Time
 	return ms.executionInfo.CloseTime.AsTime(), nil
 }
 
+// GetWorkflowExecutionDuration returns the workflow execution duration.
+// Returns zero for open workflow.
+func (ms *MutableStateImpl) GetWorkflowExecutionDuration(ctx context.Context) (time.Duration, error) {
+	closeTime, err := ms.GetWorkflowCloseTime(ctx)
```

Comment:
It seems the only reason this returns an `err` is because of this
```
// This is for backward compatible. Prior to v1.16 does not have close time in mutable state (Added by 05/21/2022).
// TODO: remove this logic when all mutable state contains close time.
```
inside `GetWorkflowCloseTime`.

Wouldn't it be safe to remove that now and always have a non-error response? It would simplify every call to `GetWorkflowExecutionDuration`.

Reply:
Maybe. cc: @yiminc @yycptt @yux0

Reply:
Main concern is that if someone is using a version older than 1.16 and upgrade to the latest version, then there will be workflows in retention but no closeTime in the mutable state. v1.16 should be old enough though.